### PR TITLE
Tests/#408/oas23 body param validation

### LINKED ISSUE #408: Test Harness: OAS2/3 Body Param Validation
**User Story**

As a developer I want to have confidence that Prism's validates request payload (body).

**ACs**
- add test cases to test harness
- add at least the cases described in the examples section- 
- if any of the test fails then mark that it is an invalid result (e.g. mark as skipped) and report a Jira bug
- link that bug to this card 
- Make sure to add the actual tests to TestRails

**Additional resources**

https://stoplightio.atlassian.net/wiki/spaces/EN/pages/52887637/Prism#Prism-MockServerFeatures (please find relevant row)

**Stuff to test oas2/oas3**
- [x] required x-www-form-urlencoded param
- [x] required x-www-form-urlencoded body
- [x] required application/json body
- [x] invalid form-data type 
- [x] invalid `enum`
- [x] form-data array (csv, pipes, multi)
- [x] form-data formats (e.g. byte)
- [x] allowEmptyValue

**oas3 related**
- [x] form-data + schema
- [ ] encoding/contentType
- [ ] encoding/headers
- [ ] encoding/allowReserved
- [ ] encoding/style
- [ ] encoding/explode

**Not in scope**

multipart/form-data

**Example tests**

Given there is an operation, that can consume requests of Content-Type CTAnd it is required that the payload has X fieldWhen I send a POST request (with Content-Type of CT) to this operationAnd the request doesn't have his X required field in the payloadThen I should get back a 400 response back

Given there is an operation, that can consume requests of Content-Type CTAnd it is required that the payload provides X fieldWhen I send a POST request (with Content-Type of CT) to this operationAnd the request does have this X required fieldThen I should get back a 2xx response back

Given there is an operation, that can consume requests of Content-Type CTAnd the operation has no required field for the payloadWhen I send a POST request (with Content-Type of CT) to this operationThen I should get back a 2xx response back

When I send a POST request (with Content-Type of CT) to an operationAnd this operation does not consume requests of Content-Type CTThen I should get back a 415 response back

**Example Specs**

**Open API 2**

Consider this rather minimal oas2 spec. It has

- POST /path operation
    - that operation takes a body
    - that body is application/json (see consumes)
    - that body requires 'x' param of string type
- PUT /path operation
   - that operation takes body
   - that body is of formData type
   - it has a required X field that can take only one of the values "one" or "two"

This is just an example.
```
swagger: '2.0'
info:
  title: Open API 2 Sink
  version: '1.0'
host: 'localhost:3000'
schemes:
  - http
produces:
  - application/json
consumes:
  - application/json
paths:
  /path:
    post:
      summary: create path
      tags: []
      responses: {}
      parameters:
        - in: body
          schema:
            type: object
            properties:
              x:
                type: string
            required:
              - x
    put:
      summary: update
      tags: []
      responses: {}
      parameters:
        - in: formData
          type: string
          name: x
          required: true
          description: ''
          enum:
            - one
            - two
definitions: {}
```
**Open API 3 **

Similarly to the previous example here is OAS3 (note the difference in formData)
```
openapi: 3.0.0
info:
  title: Open API 3 Sink
  version: '1.0'
servers:
  - url: 'http://localhost:3000'
paths:
  /path:
    post:
      summary: create
      tags: []
      responses: {}
      requestBody:
        content:
          application/json:
            schema:
              type: object
              properties:
                x:
                  type: string
              required:
                - x
    put:
      summary: update
      tags: []
      responses: {}
      requestBody:
        content:
          application/x-www-form-urlencoded:
            schema:
              type: object,
              properties:
                  name:
                    description: Updated name of the pet
                    type: string
                  status:
                    description: Updated status of the pet,
                    type: string
              required:
                - status
components:
  schemas: {}
```

### PR DESCRIPTION
**Pull request by chris-miaskowski**
## Checklist

- [x] Tests have been added (for bug fixes / features)

## What kind of change does this PR introduce?

_Tests_

## Notes

Some tests fail and I will create issues for them shortly.

Closes #408

### CODE REVIEW

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+async function instantiatePrism(fixture: string) {
+  const server = createServer({}, { components: { logger }, config: { mock: { dynamic: false } } });
+  await server.prism.load({
+    path: relative(process.cwd(), resolve(__dirname, 'fixtures', fixture)),
+  });
+  return server;
+}
+
+const expectPayload = (response: fastify.HTTPInjectResponse) => {
+  const parsed = JSON.parse(response.payload);
```
Comment (lag-of-death) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: seems like `expectPayload` only depends on `response.payload`, so there's no need to pass the whole `response` object most likely

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  return expect(parsed);
+};
+
+describe('body params validation', () => {
```
Comment (lag-of-death) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: hm.. these don't look like test-harness to me? maybe we could put these into another PR?

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+      server = await instantiatePrism(`operations-with-formdata-param.${oas}.yaml`);
+    });
+
+    describe('required parameter not in body', () => {
+      test('returns 422', async () => {
+        const response = await server.fastify.inject({
+          method: 'POST',
+          url: '/path',
+          payload: {},
+          // headers: {
```
Comment (lag-of-death) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: should it be removed now?

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    describe('parameter does not match enum criteria', () => {
+      test('returns 422 & proper validation message', async () => {
+        const response = await server.fastify.inject({
+          method: 'POST',
+          url: '/path',
+          payload: {
+            id: 'not integer',
+            status: 'somerundomestuff',
+          },
+          // headers: {
```
Comment (lag-of-death) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: should it be removed now?

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    describe('valid parameter provided', () => {
+      test('returns 200', async () => {
+        const response = await server.fastify.inject({
+          method: 'POST',
+          url: '/path',
+          payload: {
+            id: 123,
+            status: 'open',
+          },
+          // headers: {
```
Comment (lag-of-death) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: should it be removed now?

File: packages/http-server/src/__tests__/body-params-validation.spec.ts (hunk 2 above)
Comment (philsturgeon) [reply packages/http-server/src/__tests__/body-params-validation.spec.ts]: I think its ok, no need for more paperwork so long as all the tests are good.

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+  afterAll(async () => {
+    await server.fastify.close();
+  });
+
+  describe.each([['oas2'], ['oas3']])('%s with body param', oas => {
+    beforeEach(async () => {
+      server = await instantiatePrism(`operations-with-body-param.${oas}.yaml`);
+    });
+
+    describe('operation with no consumes', () => {
```
Comment (philsturgeon) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: Can we talk about this in a more generic way? oas3 does not have "consumes"

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            id: 'not integer',
+            status: 'somerundomestuff',
+          },
+          // headers: {
+          // 'content-type': 'application/x-www-form-urlencoded',
+          // },
+        });
+
+        expect(response.statusCode).toBe(422);
+        const parsed = JSON.parse(response.payload);
```
Comment (philsturgeon) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: expectPayload ?

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+async function instantiatePrism(fixture: string) {
+  const server = createServer({}, { components: { logger }, config: { mock: { dynamic: false } } });
+  await server.prism.load({
+    path: relative(process.cwd(), resolve(__dirname, 'fixtures', fixture)),
+  });
+  return server;
+}
+
+const expectPayload = (response: fastify.HTTPInjectResponse) => {
```
Comment (philsturgeon) [inline packages/http-server/src/__tests__/body-params-validation.spec.ts]: I have some concerns here.

1. Payload is a generic term which often means request body but could mean request or response, but here it always means response.
2. Should this be an expect wrapper? could it be a custom matcher? 
3. this is always JSON only, we are meant to be JSON and XML at least, especially with #456 coming up this/next sprint

**Comment by chris-miaskowski** [comment]
Thanks for reviews folks. I moved this PR to https://github.com/stoplightio/prism/pull/501 and merged with @XVincentX's graphite removal.
